frame/composable-traits/src/lending/mod.rs

```diff
@@ -97,17 +114,112 @@ impl<LiquidationStrategyId, AssetId: Copy> CreateInput<LiquidationStrategyId, As
 	}
 }
 
-#[derive(Encode, Decode, Default, TypeInfo)]
+#[derive(Encode, Decode, Default, TypeInfo, Debug)]
 pub struct MarketConfig<VaultId, AssetId, AccountId, LiquidationStrategyId> {
+	/// The owner of this market.
 	pub manager: AccountId,
-	pub borrow: VaultId,
-	pub collateral: AssetId,
+	/// The vault containing the borrow asset.
+	pub borrow_asset_vault: VaultId,
+	/// The asset being used as collateral.
+	pub collateral_asset: AssetId,
 	pub collateral_factor: MoreThanOneFixedU128,
 	pub interest_rate_model: InterestRateModel,
-	pub under_collaterized_warn_percent: Percent,
+	pub under_collateralized_warn_percent: Percent,
 	pub liquidators: Vec<LiquidationStrategyId>,
 }
 
+/// Different ways that a market can be repaid.
+// REVIEW: Perhaps add an "interest only" strategy?
+// InterestOnly
+#[derive(Encode, Decode, TypeInfo, Debug, Clone, PartialEq)]
+pub enum RepayStrategy<T> {
+	/// Attempt to repay the entirety of the remaining debt.
+	TotalDebt,
+	/// Repay the specified amount, repaying interest and principal proportionately.
+	///
+	/// # Example
+	///
+	/// ```text
+	/// principal = 90
+	/// interest = 10
+	///
+	/// total_debt_with_interest = 10 + 90
+	///                          = 100
+	///
+	/// repay = 20
+	///
+	/// new_principal = principal - ((principal / total_debt_with_interest) * repay)
+	///               = 90 - ((90 / 100) * 20)
+	///               = 72
+	///
+	/// new_interest = interest - ((interest / total_debt_with_interest) * repay)
+	///              = 10 - ((10 / 100) * 20)
+	///              = 8
+	/// ```
+	PartialAmount(T),
+}
+
+/// The total amount of debt for an account on a market, if any.
+#[derive(Encode, Decode, TypeInfo, Debug, Clone, PartialEq)]
+pub enum TotalDebtWithInterest<T> {
+	/// The account has some amount of debt on the market. Guarranteed to be non-zero.
```

Comment:
@andor0 @hussein-aitlahcen what do you think of such enums which are option for one arith result (division by zero) or maximally Result<T, ArithmethicError>.
I do not think enums are good and composable.